design-proposals/volume-migration/volume-migration.md

```diff
@@ -0,0 +1,297 @@
+
+## Motivation
+In some circumstances, the guest owner may want to have the option of switching the storage provider or class for a group of volumes used by an active VM. The VM must currently be turned off and recreated for this operation causing service disruptions and downtime.
+
+
+## Goals
+Design and develop an API to handle storage migration between PVCs.
+
+ ## Non  Goals
+*  This features is not directly aimed for live migrating the VM using RWO volumes. Hence, the API covers and refers only to the volume migration.
+*  This design doesn't cover migration between storage classes, but rather implements a basic APIs that can be used by overlaying tooling
+*  This design doesn't include the creation of the destination PVCs, but it expects the destination PVCs to exist.
+
+ ## Definition of Users
+This feature addresses for all KubeVirt users who want to move the PVCs assigned to a VM without workload interruptions.
+
+## User Stories
+
+  1. As a VM owner, the storage class of volumes used by a running VM is deprecated and I need to move the VMs usin another storage class
+  2. As a VM owner, I want to use higher performance storage class that is now available in the cluster
+  3. As a VM owner, I see my non-resizable storage device filling up and I want to move my VM to a larger storage device without interruption in order to give it more storage
+  4. As an infra admin, I need to replace the storage array due to a hardware refresh and I want to do this without stopping the VMs in order to make this infra operation transparent to VM owners.
+
+## Repos
+
+  - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+This proposal introduces a new CRD called `VolumeMigration`
+
+```golang
+type VolumeMigration struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec VolumeMigrationSpec `json:"spec"`
+
+	// +optional
+	Status *VolumeMigrationStatus `json:"status,omitempty"`
+}
+
+type VolumeMigrationList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	// +listType=atomic
+	Items []VolumeMigration `json:"items"`
+}
+
+// ReclaimPolicySourcePvc describes how the source PVC will be treated after the storage migration completes.
+// The policies follows the same behavior as the RetainPolicy for PVs: https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reclaiming
+type ReclaimPolicySourcePvc string
+
+const (
+	DeleteReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Delete"
+	RetainReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Retain"
+)
+
+type MigratedVolume struct {
+	SourcePvc      string `json:"sourcePvc,omitempty" valid:"required"`
+	DestinationPvc string `json:"destinationPvc,omitempty" valid:"required"`
+	// ReclaimPolicySourcePvc describes how the source volumes will be
+	// treated after a successful migration
+	ReclaimPolicySourcePvc ReclaimPolicySourcePvc `json:"reclaimPolicySourcePvc,omitempty"`
```

Comment:
Since this allows omitempty, what would be the behavior if the value is not specified? Would it default to retain or delete? Maybe mention what the expected behavior is in that case in a comment.

Reply:
I'll put this as a required field